File: src/skills/custom/remote-logger-implementation/scripts/remote_logger.py
import json
import random
import re
import time
from dataclasses import dataclass
from typing import Any, Dict, List, Optional
from urllib.error import HTTPError, URLError
from urllib.parse import quote
from urllib.request import Request, urlopen
import os

BASE_URL = os.environ.get("LETTA_LOGGER_API", "http://100.80.49.10:8284/libraries/local-php-api")

RUNNING_COLOR = "lightyellow"
PASS_COLOR = "lightgreen"
FAIL_COLOR = "#fb6666"
MAX_LOG_ENTRIES = 120
MAX_MESSAGE_CHARS = 800
MAX_OBJECT_DATA_BYTES = 200_000
ANSI_ESCAPE_PATTERN = re.compile(r"\x1b(?:[@-Z\\-_]|\[[0-?]*[ -/]*[@-~])")


@dataclass
class RemoteLogger:
    object_view_id: str

    def __post_init__(self) -> None:
        self.log_objects: List[Dict[str, Any]] = []
        self.monitor_led: Dict[str, Any] = default_led()

    def init(self) -> None:
        existing = self._fetch_existing_state()
        if existing is not None:
            self.log_objects = existing.get("logObjects", [])
            raw_led = existing.get("monitorLed") or default_led()
            if raw_led.get("classObject") is None:
                raw_led = {**raw_led, "classObject": default_led()["classObject"]}
                self.monitor_led = raw_led
                self._post("update")
            else:
                self.monitor_led = raw_led
            return

        self._post("insert")

    def log(self, message: str) -> None:
        safe_message = sanitize_log_message(str(message)[:MAX_MESSAGE_CHARS])
        timestamp = int(time.time() * 1000)
        rid = random.randint(0, 10**13 - 1)

        self.log_objects.append(
            {
                "timestamp": timestamp,
                "id": f"{self.object_view_id}_{rid}_{timestamp}",
                "message": safe_message,
                "method": "createLogObject",
            }
        )
        self._shrink_state_for_transport()
        self.monitor_led = updated_led(safe_message, self.monitor_led)
        self._post("update")

    def clear_logs(self, led_text: str = "ready.") -> None:
        self.monitor_led = {**default_led(), "ledText": led_text}
        self._post("update")

    def flush_logs(self, led_text: str = "ready.") -> None:
        self.log_objects = []
        self.monitor_led = {**default_led(), "ledText": led_text}
        self._post("update")

    def _state(self) -> Dict[str, Any]:
        return {
            "object_view_id": self.object_view_id,
            "logObjects": self.log_objects,
            "monitorLed": self.monitor_led,
        }

    def _serialized_state(self) -> str:
        self._shrink_state_for_transport()
        return json.dumps(self._state(), separators=(",", ":"))

    def _shrink_state_for_transport(self) -> None:
        if len(self.log_objects) > MAX_LOG_ENTRIES:
            self.log_objects = self.log_objects[-MAX_LOG_ENTRIES:]

        serialized = json.dumps(self._state(), separators=(",", ":"))
        while len(serialized.encode("utf-8")) > MAX_OBJECT_DATA_BYTES and len(self.log_objects) > 1:
            self.log_objects.pop(0)
            serialized = json.dumps(self._state(), separators=(",", ":"))

    def _post(self, action: str, timeout_sec: float = 8.0) -> None:
        payload = {
            "object_view_id": self.object_view_id,
            "object_data": self._serialized_state(),
        }

        try:
            status, body_text = http_json(
                f"{BASE_URL}/object/{action}",
                method="POST",
                payload=payload,
                timeout_sec=timeout_sec,
            )
        except Exception as err:
            if self._is_state_persisted():
                return
            raise RuntimeError(f"[RemoteLogger] {action} failed before response: {err}") from err

        if status < 200 or status >= 300:
            if action == "update" and self._try_insert_fallback(timeout_sec):
                return
            if self._is_state_persisted_with_retry():
                return
            raise RuntimeError(f"[RemoteLogger] {action} failed (HTTP {status})")

        try:
            body = json.loads(body_text) if body_text else None
        except Exception as err:
            if self._is_state_persisted_with_retry():
                return
            raise RuntimeError(f"[RemoteLogger] {action} failed (invalid JSON response): {err}") from err

        if isinstance(body, dict) and body.get("error"):
            if action == "update" and self._try_insert_fallback(timeout_sec):
                return
            if self._is_state_persisted_with_retry():
                return
            raise RuntimeError(f"[RemoteLogger] {action} error: {json.dumps(body)}")

    def _try_insert_fallback(self, timeout_sec: float = 8.0) -> bool:
        try:
            status, _ = http_json(
                f"{BASE_URL}/object/insert",
                method="POST",
                payload={
                    "object_view_id": self.object_view_id,
                    "object_data": self._serialized_state(),
                },
                timeout_sec=timeout_sec,
            )
            if status < 200 or status >= 300:
                return False
            return self._is_state_persisted()
        except Exception:
            return False

    def _fetch_existing_state(self) -> Optional[Dict[str, Any]]:
        try:
            status, body_text = http_json(
                f"{BASE_URL}/object/select?object_view_id={quote(self.object_view_id)}",
                method="GET",
                payload=None,
                timeout_sec=8.0,
            )
            if status < 200 or status >= 300:
                return None
            body = json.loads(body_text) if body_text else None
            return self._parse_select_payload(body)
        except Exception:
            return None

    def _parse_select_payload(self, body: Any) -> Optional[Dict[str, Any]]:
        if not isinstance(body, dict) or body.get("error") or not body.get("object_data"):
            return None
        try:
            parsed = json.loads(body["object_data"])
            if isinstance(parsed, dict):
                return parsed
            return None
        except Exception:
            return None

    def _is_state_persisted(self) -> bool:
        persisted = self._fetch_existing_state()
        if not persisted or persisted.get("object_view_id") != self.object_view_id:
            return False

        if not self.log_objects:
            return True

        expected_last = self.log_objects[-1].get("id")
        persisted_entries = persisted.get("logObjects") or []
        return any(entry.get("id") == expected_last for entry in persisted_entries if isinstance(entry, dict))

    def _is_state_persisted_with_retry(self, attempts: int = 4, delay_sec: float = 0.25) -> bool:
        for i in range(attempts):
            if self._is_state_persisted():
                return True
            if i < attempts - 1:
                time.sleep(delay_sec)
        return False


def default_led() -> Dict[str, Any]:
    return {
        "classObject": {
            "background_color": RUNNING_COLOR,
            "text_align": "left",
            "margin_top": "2px",
            "color": "black",
        },
        "ledText": "ready.",
        "RUNNING_COLOR": RUNNING_COLOR,
        "PASS_COLOR": PASS_COLOR,
        "FAIL_COLOR": FAIL_COLOR,
    }


def updated_led(message: str, current: Dict[str, Any]) -> Dict[str, Any]:
    cls = dict(default_led()["classObject"])
    cls.update((current.get("classObject") or {}))
    led = dict(current)
    led["classObject"] = cls
    led["ledText"] = message

    is_timeout_like = (
        re.search(r"timed?\s*out", message, re.IGNORECASE)
        or re.search(r"\btimeout\b", message, re.IGNORECASE)
        or re.search(r"\bhung\b", message, re.IGNORECASE)
        or re.search(r"\babort(?:ed|error)?\b", message, re.IGNORECASE)
    )

    if "finished" in message or re.search(r"\bPASS(?:ED)?\b", message) or re.search(r"test complete", message, re.IGNORECASE):
        led["classObject"]["background_color"] = PASS_COLOR
        led["classObject"]["color"] = "black"
    elif "ERROR" in message or re.search(r"\bFAIL(?:ED)?\b", message) or is_timeout_like:
        led["classObject"]["background_color"] = FAIL_COLOR
        led["classObject"]["color"] = "white"
    else:
        led["classObject"]["background_color"] = RUNNING_COLOR
        led["classObject"]["color"] = "black"

    return led


def sanitize_log_message(message: str) -> str:
    out = ANSI_ESCAPE_PATTERN.sub("", message)
    out = out.replace("'", "\u2019").replace('"', "\u201d")
    out = re.sub(r"[\r\t]+", " ", out)
    out = re.sub(r"[\x00-\x08\x0B\x0C\x0E-\x1F\x7F]", " ", out)
    out = re.sub(r" {2,}", " ", out)
    out = re.sub(r"\n+", " | ", out)
    return out.strip()


def http_json(url: str, method: str, payload: Optional[Dict[str, Any]], timeout_sec: float) -> (int, str):
    data = None
    headers = {}

    if payload is not None:
        data = json.dumps(payload).encode("utf-8")
        headers["Content-Type"] = "application/json"

    req = Request(url=url, method=method, data=data, headers=headers)

    try:
        with urlopen(req, timeout=timeout_sec) as res:
            body = res.read().decode("utf-8", errors="replace")
            return int(res.status), body
    except HTTPError as err:
        body = err.read().decode("utf-8", errors="replace")
        return int(err.code), body
    except URLError:
        raise


if __name__ == "__main__":
    logger = RemoteLogger("PythonExample_2026")
    logger.init()
    logger.clear_logs("python logger ready")
    logger.log("step 1")
    logger.log("PASS: finished")

File: src/skills/custom/remote-logger-implementation/scripts/remote_logger.ts
const BASE_URL =
  process.env.LETTA_LOGGER_API ??
  "http://100.80.49.10:8284/libraries/local-php-api";

type LogEntry = {
  timestamp: number;
  id: string;
  message: string;
  method: string;
};

type MonitorLedClassObject = {
  background_color: string;
  text_align: string;
  margin_top: string;
  color: string;
};

type MonitorLed = {
  classObject: MonitorLedClassObject;
  ledText: string;
  RUNNING_COLOR: string;
  PASS_COLOR: string;
  FAIL_COLOR: string;
};

type LoggerState = {
  object_view_id: string;
  logObjects: LogEntry[];
  monitorLed: MonitorLed;
};

const RUNNING_COLOR = "lightyellow";
const PASS_COLOR = "lightgreen";
const FAIL_COLOR = "#fb6666";
const MAX_LOG_ENTRIES = 120;
const MAX_MESSAGE_CHARS = 800;
const MAX_OBJECT_DATA_BYTES = 200_000;
const ANSI_ESCAPE_PATTERN = /\u001b(?:[@-Z\\-_]|\[[0-?]*[ -/]*[@-~])/g;
const SAFE_APOSTROPHE = "\u2019";
const SAFE_DOUBLE_QUOTE = "\u201d";

function defaultLed(): MonitorLed {
  return {
    classObject: {
      background_color: RUNNING_COLOR,
      text_align: "left",
      margin_top: "2px",
      color: "black",
    },
    ledText: "ready.",
    RUNNING_COLOR,
    PASS_COLOR,
    FAIL_COLOR,
  };
}

function updatedLed(message: string, current: MonitorLed): MonitorLed {
  const led: MonitorLed = {
    ...current,
    classObject: {
      ...defaultLed().classObject,
      ...(current.classObject ?? {}),
    },
    ledText: message,
  };

  const isTimeoutLike =
    /timed?\s*out/i.test(message) ||
    /\btimeout\b/i.test(message) ||
    /\bhung\b/i.test(message) ||
    /\babort(?:ed|error)?\b/i.test(message);

  if (
    message.includes("finished") ||
    /\bPASS(?:ED)?\b/.test(message) ||
    /test complete/i.test(message)
  ) {
    led.classObject.background_color = PASS_COLOR;
    led.classObject.color = "black";
  } else if (
    message.includes("ERROR") ||
    /\bFAIL(?:ED)?\b/.test(message) ||
    isTimeoutLike
  ) {
    led.classObject.background_color = FAIL_COLOR;
    led.classObject.color = "white";
  } else {
    led.classObject.background_color = RUNNING_COLOR;
    led.classObject.color = "black";
  }

  return led;
}

function sanitizeLogMessage(message: string): string {
  return message
    .replace(ANSI_ESCAPE_PATTERN, "")
    .replace(/'/g, SAFE_APOSTROPHE)
    .replace(/"/g, SAFE_DOUBLE_QUOTE)
    .replace(/[\r\t]+/g, " ")
    .replace(/[\u0000-\u0008\u000B\u000C\u000E-\u001F\u007F]/g, " ")
    .replace(/ {2,}/g, " ")
    .replace(/\n+/g, " | ")
    .trim();
}

export class RemoteLogger {
  private objectViewId: string;
  private logObjects: LogEntry[] = [];
  private monitorLed: MonitorLed = defaultLed();

  constructor(objectViewId: string) {
    this.objectViewId = objectViewId;
  }

  async init(): Promise<void> {
    const existing = await this.fetchExistingState();
    if (existing) {
      this.logObjects = existing.logObjects ?? [];
      const rawLed = existing.monitorLed ?? defaultLed();
      this.monitorLed = rawLed.classObject
        ? rawLed
        : { ...rawLed, classObject: defaultLed().classObject };
      if (!existing.monitorLed?.classObject) {
        await this.post("update");
      }
      return;
    }
    await this.post("insert");
  }

  async log(message: string): Promise<void> {
    const safeMessage = sanitizeLogMessage(
      String(message).slice(0, MAX_MESSAGE_CHARS),
    );
    const timestamp = Date.now();
    const rand = Math.floor(Math.random() * 1e13);

    this.logObjects.push({
      timestamp,
      id: `${this.objectViewId}_${rand}_${timestamp}`,
      message: safeMessage,
      method: "createLogObject",
    });

    this.shrinkStateForTransport();
    this.monitorLed = updatedLed(safeMessage, this.monitorLed);
    await this.post("update");
  }

  async clearLogs(ledText = "ready."): Promise<void> {
    this.monitorLed = { ...defaultLed(), ledText };
    await this.post("update");
  }

  async flushLogs(ledText = "ready."): Promise<void> {
    this.logObjects = [];
    this.monitorLed = { ...defaultLed(), ledText };
    await this.post("update");
  }

  private state(): LoggerState {
    return {
      object_view_id: this.objectViewId,
      logObjects: this.logObjects,
      monitorLed: this.monitorLed,
    };
  }

  private serializedState(): string {
    this.shrinkStateForTransport();
    return JSON.stringify(this.state());
  }

  private shrinkStateForTransport(): void {
    if (this.logObjects.length > MAX_LOG_ENTRIES) {
      this.logObjects = this.logObjects.slice(-MAX_LOG_ENTRIES);
    }

    let serialized = JSON.stringify(this.state());
    while (
      serialized.length > MAX_OBJECT_DATA_BYTES &&
      this.logObjects.length > 1
    ) {
      this.logObjects.shift();
      serialized = JSON.stringify(this.state());
    }
  }

  private async post(
    action: "insert" | "update",
    timeoutMs = 8000,
  ): Promise<void> {
    const payload = {
      object_view_id: this.objectViewId,
      object_data: this.serializedState(),
    };

    let res: Response;
    try {
      res = await fetchWithTimeout(
        `${BASE_URL}/object/${action}`,
        {
          method: "POST",
          headers: { "Content-Type": "application/json" },
          body: JSON.stringify(payload),
        },
        timeoutMs,
      );
    } catch (err) {
      if (await this.isStatePersisted()) return;
      throw new Error(
        `[RemoteLogger] ${action} failed before response: ${err instanceof Error ? err.message : String(err)}`,
      );
    }

    if (!res.ok) {
      if (action === "update" && (await this.tryInsertFallback(timeoutMs)))
        return;
      if (await this.isStatePersistedWithRetry()) return;
      throw new Error(`[RemoteLogger] ${action} failed (HTTP ${res.status})`);
    }

    try {
      const body = (await res.json()) as Record<string, unknown> | null;
      if (body?.error) {
        if (action === "update" && (await this.tryInsertFallback(timeoutMs)))
          return;
        if (await this.isStatePersistedWithRetry()) return;
        throw new Error(
          `[RemoteLogger] ${action} error: ${JSON.stringify(body)}`,
        );
      }
    } catch (err) {
      if (await this.isStatePersistedWithRetry()) return;
      throw new Error(
        `[RemoteLogger] ${action} failed (invalid JSON response): ${err instanceof Error ? err.message : String(err)}`,
      );
    }
  }

  private async tryInsertFallback(timeoutMs = 8000): Promise<boolean> {
    try {
      const res = await fetchWithTimeout(
        `${BASE_URL}/object/insert`,
        {
          method: "POST",
          headers: { "Content-Type": "application/json" },
          body: JSON.stringify({
            object_view_id: this.objectViewId,
            object_data: this.serializedState(),
          }),
        },
        timeoutMs,
      );
      if (!res.ok) return false;
      return this.isStatePersisted();
    } catch {
      return false;
    }
  }

  private async fetchExistingState(): Promise<LoggerState | null> {
    try {
      const res = await fetchWithTimeout(
        `${BASE_URL}/object/select?object_view_id=${encodeURIComponent(this.objectViewId)}`,
        { method: "GET" },
        8000,
      );
      if (!res.ok) return null;
      const data = (await res.json()) as Record<string, unknown> | null;
      return this.parseSelectPayload(data);
    } catch {
      return null;
    }
  }

  private parseSelectPayload(
    data: Record<string, unknown> | null,
  ): LoggerState | null {
    if (!data || data.error || !data.object_data) return null;
    try {
      return JSON.parse(data.object_data as string) as LoggerState;
    } catch {
      return null;
    }
  }

  private async isStatePersisted(): Promise<boolean> {
    const persisted = await this.fetchExistingState();
    if (!persisted || persisted.object_view_id !== this.objectViewId)
      return false;

    const expectedLast = this.logObjects[this.logObjects.length - 1];
    if (!expectedLast) return true;

    return (
      persisted.logObjects?.some((entry) => entry.id === expectedLast.id) ??
      false
    );
  }

  private async isStatePersistedWithRetry(
    attempts = 4,
    delayMs = 250,
  ): Promise<boolean> {
    for (let i = 0; i < attempts; i += 1) {
      if (await this.isStatePersisted()) return true;
      if (i < attempts - 1) {
        await new Promise((resolve) => setTimeout(resolve, delayMs));
      }
    }
    return false;
  }
}

async function fetchWithTimeout(
  url: string,
  init: RequestInit,
  timeoutMs: number,
): Promise<Response> {
  const ac = new AbortController();
  const t = setTimeout(() => ac.abort(), timeoutMs);
  try {
    return await fetch(url, { ...init, signal: ac.signal });
  } finally {
    clearTimeout(t);
  }
}
